### PR TITLE
*: make prepare stmt retriable when schema is out-dated.

### PR DESCRIPTION
**Pull request by hanfei1991**
We should use ast node to try instead of raw stmt. Because execute stmt will change its original text for printing log, it's unreasonable !

### CODE REVIEW

File: plan/optimizer.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/privilege"
 	"github.com/pingcap/tidb/terror"
+	"github.com/ngaut/log"
```
Comment (shenli) [inline plan/optimizer.go]: We use another log lib.

File: plan/optimizer.go [hunk 2]
```diff
@@ -190,6 +191,7 @@ func PrepareStmt(is infoschema.InfoSchema, ctx context.Context, node ast.Node) e
 	if err := Preprocess(node, is, ctx); err != nil {
 		return errors.Trace(err)
 	}
+	log.Warnf("validate")
```
Comment (shenli) [inline plan/optimizer.go]: Remove this?

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM
@coocood PTAL

**Comment by tiancaiamao** [comment]
> "execute stmt will change its original text for printing log"

That's not true!

**Comment by tiancaiamao** [comment]
LGTM @coocood

File: executor/adapter.go [hunk 3]
```diff
@@ -93,20 +93,20 @@ func (a *recordSet) Close() error {
...
-	plan           plan.Plan
-	startTime      time.Time
-	isPreparedStmt bool
-	expensive      bool
+	ctx       context.Context
+	text      string
+	plan      plan.Plan
+	startTime time.Time
+	expensive bool
+	astNodes  ast.StmtNode
```
Comment (zz-jason) [inline executor/adapter.go]: s/astNodes/astNode/

File: ast/ast.go [hunk 4]
```diff
@@ -174,8 +174,8 @@ type Statement interface {
 	// Exec executes SQL and gets a Recordset.
 	Exec(ctx context.Context) (RecordSet, error)
 
-	// IsPrepared returns whether this statement is prepared statement.
-	IsPrepared() bool
+	// AstNodes return the ast nodes for retry.
+	AstNodes() StmtNode
```
Comment (zz-jason) [inline ast/ast.go]: s/AstNodes/AstNode/

**Pull request merged**